File: java/src/com/mogreet/mercury/user/Getopt.java
/**
* Copyright 2011 Nicolas Payot
*
* This file is part of Mercury.
*
* Mercury is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Mercury is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Mercury.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.mogreet.mercury.user;

import com.mogreet.mercury.*;
import java.util.*;
import org.w3c.dom.*;
import javax.xml.xpath.*;
import java.lang.Math.*;

/**
* The Getopt object contains the response from a {@link Mercury#getopt} request to the Moms API.
*/
public class Getopt extends Response {
	
	private HashMap<Integer, HashMap<String, String>> campaigns;
	
	/**
	* Creates a new Getopt response object.
	* @param xmlDoc XML DOM document that is returned by all the requests.
	*/
	public Getopt(Document xmlDoc) {
		super(xmlDoc);
		if (!super.responseIsValid())
			throw new RuntimeException(super.getMessage());
		try {
			XPath xpath = XPathFactory.newInstance().newXPath();
			XPathExpression expr = xpath.compile("/response/campaign/@id | /response/campaign/status/@code | /response/campaign/status/text()");
			
			NodeList nodes = (NodeList)expr.evaluate(xmlDoc, XPathConstants.NODESET);
			this.campaigns = new HashMap<Integer, HashMap<String, String>>();
			int campaignId = 0;
			for (int i = 0; i < nodes.getLength(); ++i) {
				if (nodes.item(i).getNodeName().equals("id")) {
					campaignId = Integer.parseInt(nodes.item(i).getNodeValue());
					this.campaigns.put(campaignId, new HashMap<String, String>());
				}	
				else {
					HashMap<String, String> status = this.campaigns.get(campaignId);
					if (nodes.item(i).getNodeName().equals("code"))
						status.put("code", nodes.item(i).getNodeValue());
					else 
						status.put("status", nodes.item(i).getNodeValue());
				}
			}
		} catch (XPathExpressionException e) {
			throw new RuntimeException("\nAn error occured while parsing the XML data for the GETOPT call: " + e.getMessage());
		}
	}
	
	/**
	* @return The list of all the campaigns id.
	*/
	public ArrayList<Integer> getCampaignsIdList() {
		ArrayList<Integer> campaignsIdList = new ArrayList<Integer>(); 
		campaignsIdList.addAll(this.campaigns.keySet());
		return campaignsIdList;
	}
	
	/**
	* @param campaignId campaign ID
	* @return The status code of the specified campaign.
	* 0 if the specified campaign ID does not exist.
	*/
	public int getCampaignStatusCode(int campaignId) {
		HashMap<String, String> campaign = this.campaigns.get(campaignId);
		if (campaign != null)
			return Integer.parseInt(campaign.get("code"));
		return 0;
	}
	
	/**
	* @param campaignId campaign ID
	* @return The status label of the specified campaign.
	* Null if the specified campaign ID does not exist.
	*/
	public String getCampaignStatus(int campaignId) {
		HashMap<String, String> campaign = this.campaigns.get(campaignId);
		if (campaign != null)
			return campaign.get("status");
		return null;
	}
}

File: dotNet/MogreetSDK/MogreetSDK/user/Getopt.cs
using System;
using System.Xml;
using System.Security;
using System.Collections.Generic;

namespace MogreetSDK.user
{
    /// <summary>
    /// The Getopt object contains the response from a <see cref="MogreetSDK.Mercury.getopt"/> request to the Moms API.
    /// </summary>
    /// <exception cref='Exception'>
    /// Represents errors that occur during application execution.
    /// </exception>
    public class Getopt : Response
    {
        private Dictionary<int, Dictionary<string, string>> campaigns;
        private int campaignId;
        private string code;
        private string status = "";

        /// <summary>
        /// Initializes a new instance of the <see cref="SDK.Getopt"/> class.
        /// </summary>
        /// <param name='xmlDoc'>
        /// Xml document that is returned by all the requests.
        /// </param>
        /// <exception cref='Exception'>
        /// Represents errors that occur during application execution.
        /// </exception>
        public Getopt(XmlDocument xmlDoc)
            : base(xmlDoc)
        {
            if (!(base.responseIsValid()))
                throw new Exception(base.getMessage());
            try
            {

                XmlNodeList nodes = mXmlDoc.SelectNodes("//response/campaign");
                this.campaigns = new Dictionary<int, Dictionary<string, string>>();

                for (int i = 0; i < nodes.Count; i++)
                {
                    //recovers attributes
                    campaignId = Int32.Parse((nodes.Item(i).Attributes.Item(0)).InnerText);
                    code = (nodes.Item(i).FirstChild.Attributes.Item(0)).Value;
                    status = nodes.Item(i).InnerText;

                    //inserts elements in Dictionarys
                    Dictionary<string, string> campaignsDic = new Dictionary<string, string>();
                    campaignsDic.Add(code, status);
                    this.campaigns.Add(campaignId, campaignsDic);

                    //prints elements
                    Console.WriteLine("campaignId: " + campaignId);
                    Console.WriteLine("code: " + code);
                    Console.WriteLine("status: " + status);
                }

                getCampaignIdList();
                Console.WriteLine(getCampaignStatusCode(10655));
                Console.WriteLine(getCampaignStatus(10655));


            }
            catch (XmlException e)
            {
                throw new Exception("\nAn error occured while parsing the XML data for the GETOPT call: " + e.Message);
            }
        }

        /// <summary>
        /// Gets the list of all the campaigns id.
        /// </summary>
        /// <returns>
        /// The list of all the campaigns id.
        /// </returns>
        public List<int> getCampaignIdList()
        {
            List<int> campaignsIdList = new List<int>(this.campaigns.Keys);
            return campaignsIdList;
        }

        /// <summary>
        /// Gets the campaign status code.
        /// </summary>
        /// <returns>
        /// The status code of the specified campaign.
        /// 0 if the specified campaign ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        public int getCampaignStatusCode(int campaignId)
        {
            if (this.campaigns.ContainsKey(campaignId))
            {
                foreach (var pair in campaigns[campaignId])
                    return Int32.Parse(pair.Key);
            }
            return 0;
        }

        /// <summary>
        /// Gets the campaign status.
        /// </summary>
        /// <returns>
        /// The status label of the specified campaign.
        /// Null if the specified campaign ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        public string getCampaignStatus(int campaignId)
        {
            if (this.campaigns.ContainsKey(campaignId))
            {
                foreach (var pair in campaigns[campaignId])
                    return pair.Value;
            }
            return null;
        }
    }
}
